tests/bgp/test_bgp_peer_shutdown.py

```diff
@@ -0,0 +1,208 @@
+    packets = sniff(
+        offline=pcap_file,
+        lfilter=lambda p: IP in p and bgp.BGPHeader in p and p[bgp.BGPHeader].type == 3,
+    )
+    return packets
+
+
+def match_bgp_notification(packet, src_ip, dst_ip, action):
+    """Check if the bgp notification packet matches."""
+    if not (packet[IP].src == src_ip and packet[IP].dst == dst_ip):
+        return False
+
+    bgp_fields = packet[bgp.BGPNotification].fields
+    if action == "cease":
+        # error_code 6: Cease, error_subcode 3: Peer De-configured. References: RFC 4271
+        return bgp_fields["error_code"] == 6 and bgp_fields["error_subcode"] == 3
+    else:
+        return False
+
+
+def is_neighbor_session_down(duthost, neighbor):
+    # handle both multi-asic and single-asic
+    bgp_neighbors = duthost.bgp_facts(num_npus=duthost.sonichost.num_asics())["ansible_facts"]["bgp_neighbors"]
+    return (neighbor.ip in bgp_neighbors and
+            bgp_neighbors[neighbor.ip]["admin"] == "down" and
+            bgp_neighbors[neighbor.ip]["state"] == "idle")
+
+
+def test_bgp_peer_shutdown(
+    common_setup_teardown,
+    constants,
+    duthosts,
+    enum_rand_one_per_hwsku_frontend_hostname,
+    request,
+):
+    duthost = duthosts[enum_rand_one_per_hwsku_frontend_hostname]
+    n0 = common_setup_teardown
+    announced_route = {"prefix": "10.10.100.0/27", "nexthop": n0.ip}
+
+    try:
+        n0.start_session()
+        # ensure new session is ready
+        if not wait_until(
+            WAIT_TIMEOUT,
+            5,
+            20,
+            lambda: is_neighbor_session_established(duthost, n0),
+        ):
+            pytest.fail("Could not establish bgp sessions")
+
+        n0.announce_route(announced_route)
+        time.sleep(constants.sleep_interval)
+        announced_route_on_dut_before_shutdown = duthost.get_route(announced_route["prefix"], n0.namespace)
+        if not announced_route_on_dut_before_shutdown:
+            pytest.fail("announce route %s from n0 to dut failed" % announced_route["prefix"])
+
+        # Tear down BGP session on n0
+        bgp_pcap = BGP_DOWN_LOG_TMPL
+        with capture_bgp_packages_to_file(duthost, "any", bgp_pcap, n0.namespace):
+            n0.teardown_session()
+            if not wait_until(
+                WAIT_TIMEOUT,
+                5,
+                20,
+                lambda: is_neighbor_session_down(duthost, n0),
+            ):
+                pytest.fail("Could not tear down bgp session")
+
+        local_pcap_filename = fetch_and_delete_pcap_file(bgp_pcap, constants.log_dir, duthost, request)
+        bpg_notifications = bgp_notification_packets(local_pcap_filename)
+        for bgp_packet in bpg_notifications:
+            logging.debug(
+                "bgp notification packet, capture time %s, packet details:\n%s",
+                bgp_packet.time,
+                bgp_packet.show(dump=True),
+            )
+
+            if not match_bgp_notification(bgp_packet, n0.ip, n0.peer_ip, "cease"):
```

Comment:
hi @cyw233 

1. should we compare the timestamp with when this session actually goes down? Receive cease pkt first then the session goes down, it would be expected behavior. 
2. could we repeat, bgp shutdown / bgp startup for several times to see if it doesn't cause any issue? maybe 5?